### PR TITLE
[SYCL-MLIR] Drop SYCL3DGridOps returning an index

### PR DESCRIPTION
**Pull request by victor-eds**
SYCL3DGridOps come in two flavors:

- One returning an `index` and receiving an `i32` argument (the component of the 3D grid element to query);
- One returning an `sycl.(id|range)<(1|2|3)>`, the 3D grid element itself.

The first flavor is dropped in this commit as:

- It does not represent any SYCL construct: whereas the second represents how the `id`/`item`/`nd_item` input to a kernel is built, the one returning an `index` does not represent any SYCL construct.
- It cannot be lowered to LLVM: Due to the OpenCL-SYCL different unit strides conventions, we would need to perform a mirroring of the components when lowering. This mirroring depends on the number of dimensions of the kernel. As we do not have that information attached to the operation, we cannot perform the mirroring, so the lowering cannot occur.

This commit also drops the traits used to verify the type of the operations, as this is no longer needed due to the introduced simplicity of the operations.

### CODE REVIEW

**Comment by victor-eds** [comment]
We are not using this operation and I don't think we'll use it, so this does not imply changes in `cgeist`

File: mlir-sycl/lib/Conversion/SYCLToSPIRV/SYCLToSPIRV.cpp [hunk 1]
```diff
@@ -294,9 +246,7 @@ class SingleDimGridOpPattern : public GridOpPattern<OpTy> {
 template <typename... OpTys>
 void addGridOpPatterns(RewritePatternSet &patterns,
                        TypeConverter &typeConverter, MLIRContext *context) {
-  (patterns.add<GridOpPatternIndex<OpTys>, GridOpPatternNoIndex<OpTys>>(
-       typeConverter, context),
-   ...);
+  (patterns.add<GridOpPatternNoIndex<OpTys>>(typeConverter, context), ...);
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Conversion/SYCLToSPIRV/SYCLToSPIRV.cpp:249]: As now we only have one pattern, should we just remove addGridOpPatterns? and do something like:
```
  patterns.add<GridOpPatternNoIndex<SYCLGlobalOffsetOp>,
               GridOpPatternNoIndex<SYCLNumWorkGroupsOp>,
               .......>(typeConverter, context);
```

File: mlir-sycl/lib/Conversion/SYCLToGPU/SYCLToGPU.cpp [hunk 2]
```diff
@@ -243,9 +196,7 @@ class SingleDimGridOpPattern final : public GridOpPattern<OpTy> {
 
 template <typename... OpTys>
 void addGridOpPatterns(RewritePatternSet &patterns, MLIRContext *context) {
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Conversion/SYCLToGPU/SYCLToGPU.cpp:198]: inline `addGridOpPatterns`?

**Review by whitneywhtsang — APPROVED** [review]

File: mlir-sycl/lib/Conversion/SYCLToSPIRV/SYCLToSPIRV.cpp (hunk 1 above)
Comment (victor-eds) [reply mlir-sycl/lib/Conversion/SYCLToSPIRV/SYCLToSPIRV.cpp]: ue. Done that and renamed the pattern to make more sense.

**Pull request merged**